File: expo-mediastore/ios/Sources/ExpoMediastore/MediaStoreModule.swift
import ExpoModulesCore

public class MediaStoreModule: Module {
  public func definition() -> ModuleDefinition {
    Name("MediaStore")

    Events("onMediaChange")

    AsyncFunction("getAudio") { (sort: [String: Any]?, filter: [String: Any]?, pagination: [String: Any]?) -> [[String: Any]] in
      return []
    }

    AsyncFunction("getVideos") { (sort: [String: Any]?, filter: [String: Any]?, pagination: [String: Any]?) -> [[String: Any]] in
      return []
    }

    AsyncFunction("getImages") { (sort: [String: Any]?, filter: [String: Any]?, pagination: [String: Any]?) -> [[String: Any]] in
      return []
    }

    AsyncFunction("getDocuments") { (sort: [String: Any]?, filter: [String: Any]?, pagination: [String: Any]?) -> [[String: Any]] in
      return []
    }

    AsyncFunction("getAlbums") { (sort: [String: Any]?, filter: [String: Any]?, pagination: [String: Any]?) -> [[String: Any]] in
      return []
    }

    AsyncFunction("getArtists") { (sort: [String: Any]?, pagination: [String: Any]?) -> [[String: Any]] in
      return []
    }

    AsyncFunction("getGenres") { (sort: [String: Any]?, pagination: [String: Any]?) -> [[String: Any]] in
      return []
    }

    AsyncFunction("getPlaylists") { (sort: [String: Any]?, pagination: [String: Any]?) -> [[String: Any]] in
      return []
    }

    AsyncFunction("getFolders") { (sort: [String: Any]?, filter: [String: Any]?, pagination: [String: Any]?) -> [[String: Any]] in
      return []
    }

    AsyncFunction("search") { (options: [String: Any]) -> [String: Any] in
      return [
        "audio": [],
        "videos": [],
        "images": [],
        "documents": [],
        "totalCount": 0,
        "query": options["query"] as? String ?? ""
      ]
    }

    AsyncFunction("getById") { (mediaType: String, id: String) -> [String: Any]? in
      return nil
    }

    AsyncFunction("getByUri") { (uri: String) -> [String: Any]? in
      return nil
    }

    AsyncFunction("getRecent") { (mediaType: String?, limit: Int?) -> [[String: Any]] in
      return []
    }

    AsyncFunction("getFavorites") { (mediaType: String?, sort: [String: Any]?, pagination: [String: Any]?) -> [[String: Any]] in
      return []
    }

    AsyncFunction("getLargestFiles") { (mediaType: String?, limit: Int?) -> [[String: Any]] in
      return []
    }

    AsyncFunction("getDuplicates") { (mediaType: String?) -> [[String: Any]] in
      return []
    }

    AsyncFunction("getStatistics") { () -> [String: Any] in
      return [
        "totalAudio": 0,
        "totalVideo": 0,
        "totalImages": 0,
        "totalDocuments": 0,
        "totalSize": 0,
        "totalDuration": 0
      ]
    }

    AsyncFunction("refresh") { () -> Void in
    }

    Function("checkPermissions") { () -> [String: Any] in
      return [
        "granted": false,
        "audio": false,
        "video": false,
        "images": false
      ]
    }

    AsyncFunction("requestPermissions") { () -> [String: Any] in
      return [
        "granted": false,
        "audio": false,
        "video": false,
        "images": false
      ]
    }
  }
}

File: src/MediaStoreModule.ts
import type { TurboModule } from "react-native";
import { TurboModuleRegistry } from "react-native";
import type {
  AudioItem,
  VideoItem,
  ImageItem,
  DocumentItem,
  Album,
  Artist,
  Genre,
  Playlist,
  Folder,
  FolderStatistics,
  SearchResult,
  SearchOptions,
  SortOptions,
  FilterOptions,
  PaginationOptions,
  MediaStoreStatistics,
  DuplicateItem,
  PermissionStatus,
  LibraryResult,
  IncrementalChanges,
} from "./MediaStoreModule.types";
import type { DetailedMetadata, MediaMetaType } from "./metadata.types";

export interface Spec extends TurboModule {
  getAudio(sort: SortOptions | null, filter: FilterOptions | null, pagination: PaginationOptions | null): Promise<AudioItem[]>;
  getVideos(sort: SortOptions | null, filter: FilterOptions | null, pagination: PaginationOptions | null): Promise<VideoItem[]>;
  getImages(sort: SortOptions | null, filter: FilterOptions | null, pagination: PaginationOptions | null): Promise<ImageItem[]>;
  getDocuments(sort: SortOptions | null, filter: FilterOptions | null, pagination: PaginationOptions | null): Promise<DocumentItem[]>;
  getAlbums(sort: SortOptions | null, filter: FilterOptions | null, pagination: PaginationOptions | null): Promise<Album[]>;
  getArtists(sort: SortOptions | null, pagination: PaginationOptions | null): Promise<Artist[]>;
  getGenres(sort: SortOptions | null, pagination: PaginationOptions | null): Promise<Genre[]>;
  getPlaylists(sort: SortOptions | null, pagination: PaginationOptions | null): Promise<Playlist[]>;
  getFolders(sort: SortOptions | null, filter: FilterOptions | null, pagination: PaginationOptions | null): Promise<Folder[]>;
  getFolderStatistics(folderPath: string | null): Promise<FolderStatistics[]>;
  getStatistics(): Promise<MediaStoreStatistics>;
  search(options: SearchOptions): Promise<SearchResult>;
  getById(mediaType: string, id: string): Promise<AudioItem | VideoItem | ImageItem | DocumentItem | null>;
  getByUri(uri: string): Promise<AudioItem | VideoItem | ImageItem | DocumentItem | null>;
  getRecent(mediaType: string | null, limit: number | null): Promise<(AudioItem | VideoItem | ImageItem | DocumentItem)[]>;
  getFavorites(mediaType: string | null, sort: SortOptions | null, pagination: PaginationOptions | null): Promise<(AudioItem | VideoItem | ImageItem | DocumentItem)[]>;
  getLargestFiles(mediaType: string | null, limit: number | null): Promise<(AudioItem | VideoItem | ImageItem | DocumentItem)[]>;
  getDuplicates(mediaType: string | null): Promise<DuplicateItem[]>;
  refresh(): Promise<void>;
  refreshIncremental(lastTimestamp: number | null): Promise<IncrementalChanges>;
  getLastRefreshTimestamp(): Promise<number>;
  checkPermissions(): Promise<PermissionStatus>;
  requestPermissions(): Promise<PermissionStatus>;
  getAlbumArtwork(albumId: string): Promise<string | null>;
  getVideoThumbnail(videoId: string, width: number | null, height: number | null): Promise<string | null>;
  getImageThumbnail(imageId: string, width: number | null, height: number | null): Promise<string | null>;
  getLibrary(sort: SortOptions | null, filter: FilterOptions | null, pagination: PaginationOptions | null): Promise<LibraryResult>;
  getDetailedMetadata(mediaType: string, id: string): Promise<DetailedMetadata | null>;
  getDetailedMetadataByUri(uri: string): Promise<DetailedMetadata | null>;
  addListener(eventName: string): void;
  removeListeners(count: number): void;
}

export default TurboModuleRegistry.getEnforcing<Spec>("MediaStore");
